### PR TITLE
[10341] Update performance Latency test from Fast-RTPS to Fast-DDS API

### PR DESCRIPTION
**Pull request by MiguelBarro**
Evaluate if the new API introduces any performance hit.
Allow a future pull request that extends the testing to new Zero Copy feature.
This PR should be merged after PR #1706.
This PR should be merged after PR #1730.

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3624/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/810/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3203/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3652/)

File: test/performance/latency/LatencyTestPublisher.cpp [hunk 1]
```diff
@@ -177,409 +199,466 @@ bool LatencyTestPublisher::init(
...
-    while (discovery_count_ != (subscribers_ * 4))
-    {
-        discovery_cv_.wait(disc_lock);
-    }
+    unique_lock<mutex> disc_lock(mutex_);
+    discovery_cv_.wait(
+        disc_lock,
+        [this]() -> bool
+        {
+            return total_matches() == 4;
```
Comment (MiguelCompany) [inline test/performance/latency/LatencyTestPublisher.cpp:641]: There could be more than one latency subscriber running...

```suggestion
            return total_matches() == 4 * subscribers_;
```

File: test/performance/latency/LatencyTestPublisher.cpp [hunk 2]
```diff
@@ -820,12 +915,28 @@ void LatencyTestPublisher::print_stats(
...
+    // condition variable wait predicate
+
+    int32_t count = data_writer_listener_.matched_
+            + data_reader_listener_.matched_
+            + command_writer_listener_.matched_
+            + command_reader_listener_.matched_;
+
+    // Each endpoint has a mirror counterpart in the LatencyTestPublisher
+    // thus, the maximun number of matches is 4
+    assert(count >= 0 && count < 5);
```
Comment (MiguelCompany) [inline test/performance/latency/LatencyTestPublisher.cpp:940]: This assumption is false. The maximum is `4 * subscribers_`

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
Partial review: LatencyTestPublisher only.

File: test/performance/latency/LatencyTestSubscriber.cpp [hunk 3]
```diff
@@ -461,50 +588,80 @@ bool LatencyTestSubscriber::test(
...
-                return command_msg_count_ > 0;
-            });
-    --command_msg_count_;
+    command_msg_cv_.wait(
+        lock,
+        [this]()
+        {
+            return command_msg_count_ != 0;
+        });
+    command_msg_count_ = 0;
```
Comment (MiguelCompany) [inline test/performance/latency/LatencyTestSubscriber.cpp:627]:
```suggestion
    --command_msg_count_;
```

File: test/performance/latency/LatencyTestSubscriber.cpp [hunk 4]
```diff
@@ -461,50 +588,80 @@ bool LatencyTestSubscriber::test(
...
 
     if (dynamic_data_)
     {
-        DynamicDataFactory::get_instance()->delete_data(dynamic_data_type_);
+        dynamic_pub_sub_type_->deleteData(dynamic_data_type_);
     }
     else
     {
-        delete(latency_type_);
+        latency_data_type_->deleteData(latency_type_);
```
Comment (MiguelCompany) [inline test/performance/latency/LatencyTestSubscriber.cpp:642]: I think the original code was ok.

File: test/performance/latency/LatencyTestSubscriber.hpp [hunk 5]
```diff
@@ -33,6 +33,14 @@
 #include <fastrtps/types/DynamicData.h>
 #include <fastrtps/types/DynamicPubSubType.h>
 
+#include <fastdds/dds/topic/TypeSupport.hpp>
```
Comment (MiguelCompany) [inline test/performance/latency/LatencyTestSubscriber.hpp:36]: Include order.

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
Review completed

File: test/performance/latency/LatencyTestPublisher.cpp [hunk 6]
```diff
@@ -21,6 +21,12 @@
 #include <fastdds/dds/log/Log.hpp>
 #include <fastdds/dds/log/Colors.hpp>
 #include <fastrtps/xmlparser/XMLProfileManager.h>
+#include <fastdds/dds/domain/DomainParticipantFactory.hpp>
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply test/performance/latency/LatencyTestPublisher.cpp]: Done in 38d969f4b12f32e97ec60f8b44d02e591713638e

File: test/performance/latency/LatencyTestPublisher.cpp [hunk 7]
```diff
@@ -177,409 +199,466 @@ bool LatencyTestPublisher::init(
...
+        }
+        else
+        {
+            ReliabilityQosPolicy rp;
+            rp.kind = eprosima::fastrtps::BEST_EFFORT_RELIABILITY_QOS;
+            dw_qos.reliability(rp);
+        }
+
+        dw_qos.properties(property_policy);
+        dw_qos.endpoint().history_memory_policy = MemoryManagementPolicy::PREALLOCATED_WITH_REALLOC_MEMORY_MODE;
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply test/performance/latency/LatencyTestPublisher.cpp]: Done in 38d969f4b12f32e97ec60f8b44d02e591713638e

File: test/performance/latency/LatencyTestPublisher.cpp [hunk 8]
```diff
@@ -177,409 +199,466 @@ bool LatencyTestPublisher::init(
...
+
+        if (reliable)
+        {
+            ReliabilityQosPolicy rp;
+            rp.kind = eprosima::fastrtps::RELIABLE_RELIABILITY_QOS;
+            dr_qos.reliability(rp);
+        }
+
+        dr_qos.properties(property_policy);
+        dr_qos.endpoint().history_memory_policy = MemoryManagementPolicy::PREALLOCATED_WITH_REALLOC_MEMORY_MODE;
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply test/performance/latency/LatencyTestPublisher.cpp]: Done in 38d969f4b12f32e97ec60f8b44d02e591713638e

File: test/performance/latency/LatencyTestPublisher.cpp (hunk 1 above)
Comment (JLBuenoLopez-eProsima) [reply test/performance/latency/LatencyTestPublisher.cpp]: Done in 38d969f4b12f32e97ec60f8b44d02e591713638e

File: test/performance/latency/LatencyTestPublisher.cpp (hunk 2 above)
Comment (JLBuenoLopez-eProsima) [reply test/performance/latency/LatencyTestPublisher.cpp]: Done in 38d969f4b12f32e97ec60f8b44d02e591713638e

File: test/performance/latency/LatencyTestPublisher.cpp [hunk 9]
```diff
@@ -177,409 +199,466 @@ bool LatencyTestPublisher::init(
...
+        DataWriter* writer,
+        const PublicationMatchedStatus& info)
 {
-    std::unique_lock<std::mutex> lock(latency_publisher_->mutex_);
+    (void)writer;
 
-    if (info.status == MATCHED_MATCHING)
-    {
-        std::cout << C_MAGENTA << "Data Pub Matched" << C_DEF << std::endl;
+    lock_guard<mutex> lock(latency_publisher_->mutex_);
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply test/performance/latency/LatencyTestPublisher.cpp]: This specific lock has been reverted in commit 38d969f4b12f32e97ec60f8b44d02e591713638e.

The other locks have been reverted in commit 4242b40554eafc23d03cffaf4df8025e529cb0f0.

File: test/performance/latency/LatencyTestSubscriber.hpp (hunk 5 above)
Comment (JLBuenoLopez-eProsima) [reply test/performance/latency/LatencyTestSubscriber.hpp]: Done in 82bfe8959317b5e4404ad4d5bce0eac19c2ebef7

File: test/performance/latency/LatencyTestPublisher.hpp [hunk 10]
```diff
@@ -37,6 +37,14 @@
 #include <fastrtps/types/DynamicData.h>
 #include <fastrtps/types/DynamicPubSubType.h>
 
+#include <fastdds/dds/topic/TypeSupport.hpp>
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply test/performance/latency/LatencyTestPublisher.hpp]: Done in 82bfe8959317b5e4404ad4d5bce0eac19c2ebef7

File: test/performance/latency/LatencyTestSubscriber.cpp [hunk 11]
```diff
@@ -16,45 +16,68 @@
  * @file LatencyTestSubscriber.cpp
  *
  */
+#include <cassert>
 
 #include "LatencyTestSubscriber.hpp"
 #include <fastdds/dds/log/Log.hpp>
 #include <fastdds/dds/log/Colors.hpp>
+#include <dds/core/status/Status.hpp>
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply test/performance/latency/LatencyTestSubscriber.cpp]: Done in f2b39f966f52e5b4c9d9b470e07a00c873e79a01

File: test/performance/latency/LatencyTestSubscriber.cpp [hunk 12]
```diff
@@ -16,45 +16,68 @@
...
 #include <fastrtps/xmlparser/XMLProfileManager.h>
+#include <fastdds/dds/domain/DomainParticipantFactory.hpp>
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply test/performance/latency/LatencyTestSubscriber.cpp]: Done in f2b39f966f52e5b4c9d9b470e07a00c873e79a01

File: test/performance/latency/LatencyTestSubscriber.cpp [hunk 13]
```diff
@@ -142,300 +168,401 @@ bool LatencyTestSubscriber::init(
...
+        }
+        else
+        {
+            ReliabilityQosPolicy rp;
+            rp.kind = eprosima::fastrtps::BEST_EFFORT_RELIABILITY_QOS;
+            dw_qos.reliability(rp);
+        }
+
+        dw_qos.properties(property_policy);
+        dw_qos.endpoint().history_memory_policy = MemoryManagementPolicy::PREALLOCATED_WITH_REALLOC_MEMORY_MODE;
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply test/performance/latency/LatencyTestSubscriber.cpp]: Done in f2b39f966f52e5b4c9d9b470e07a00c873e79a01

File: test/performance/latency/LatencyTestSubscriber.cpp [hunk 14]
```diff
@@ -142,300 +168,401 @@ bool LatencyTestSubscriber::init(
...
-    std::unique_lock<std::mutex> lock(latency_publisher_->mutex_);
+    (void)writer;
 
-    if (info.status == MATCHED_MATCHING)
-    {
-        logInfo(LatencyTest, "Data Pub Matched");
-        ++latency_publisher_->discovery_count_;
-    }
-    else
+    lock_guard<mutex> lock(latency_subscriber_->mutex_);
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply test/performance/latency/LatencyTestSubscriber.cpp]: Reverted all lock changes in commit f2b39f966f52e5b4c9d9b470e07a00c873e79a01

File: test/performance/latency/LatencyTestSubscriber.cpp [hunk 15]
```diff
@@ -142,300 +168,401 @@ bool LatencyTestSubscriber::init(
...
-            latency_publisher_->mutex_.unlock();
-            latency_publisher_->command_msg_cv_.notify_one();
+            case READY:
+                log << "Publisher has new test ready...";
+                break;
+            case STOP:
+                log << "Publisher has stopped the test";
+                break;
+            case STOP_ERROR:
+                log << "Publisher has canceled the test";
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply test/performance/latency/LatencyTestSubscriber.cpp:503]: Good catch!

Done in f2b39f966f52e5b4c9d9b470e07a00c873e79a01

File: test/performance/latency/LatencyTestSubscriber.cpp [hunk 16]
```diff
@@ -461,50 +588,80 @@ bool LatencyTestSubscriber::test(
...
-    --command_msg_count_;
+    // Wait for the Publisher READY command
+    unique_lock<mutex> lock(mutex_);
+    command_msg_cv_.wait(
+        lock,
+        [this]()
+        {
+            return command_msg_count_ != 0;
+        });
+    command_msg_count_ = 0;
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply test/performance/latency/LatencyTestSubscriber.cpp]: Done in f2b39f966f52e5b4c9d9b470e07a00c873e79a01

File: test/performance/latency/LatencyTestSubscriber.cpp (hunk 3 above)
Comment (JLBuenoLopez-eProsima) [reply test/performance/latency/LatencyTestSubscriber.cpp]: Done in f2b39f966f52e5b4c9d9b470e07a00c873e79a01

File: test/performance/latency/LatencyTestSubscriber.cpp [hunk 17]
```diff
@@ -142,300 +168,401 @@ bool LatencyTestSubscriber::init(
...
+
+        if (reliable)
+        {
+            ReliabilityQosPolicy rp;
+            rp.kind = eprosima::fastrtps::RELIABLE_RELIABILITY_QOS;
+            dr_qos.reliability(rp);
+        }
+
+        dr_qos.properties(property_policy);
+        dr_qos.endpoint().history_memory_policy = MemoryManagementPolicy::PREALLOCATED_WITH_REALLOC_MEMORY_MODE;
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply test/performance/latency/LatencyTestSubscriber.cpp]: Done in f2b39f966f52e5b4c9d9b470e07a00c873e79a01

File: test/performance/latency/LatencyTestTypes.hpp [hunk 18]
```diff
@@ -72,39 +81,30 @@ class LatencyType
...
         m_typeSize = MAX_TYPE_SIZE;
         m_isGetKeyDefined = false;
     }
 
+    LatencyDataType(
+            const uint32_t& size)
+        : buffer_size_(size)
+    {
+        setName("LatencyType");
+        m_typeSize = size + 4;
```
Comment (MiguelCompany) [inline test/performance/latency/LatencyTestTypes.hpp:104]: Use aligned size here:

```suggestion
        m_typeSize = 4 + ((size + 3) & ~3);
```

File: test/performance/latency/LatencyTestTypes.cpp [hunk 19]
```diff
@@ -24,55 +24,58 @@
...
-
         return size;
     };
 }
 
 void* LatencyDataType::createData()
 {
-
-    return (void*)new LatencyType();
+    return (void*)new uint8_t[(4 + buffer_size_ + 3) / 4 * 4];
```
Comment (MiguelCompany) [inline test/performance/latency/LatencyTestTypes.cpp:71]:
```suggestion
    return (void*)new uint8_t[m_typeSize];
```

File: test/performance/latency/LatencyTestTypes.hpp [hunk 20]
```diff
@@ -129,6 +129,27 @@ class LatencyDataType : public eprosima::fastdds::dds::TopicDataType
         return false;
     }
 
+    void set_size(
```
Comment (MiguelCompany) [inline test/performance/latency/LatencyTestTypes.hpp:132]: This method has been declared, but I see no implementation.

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
Partial review: data type

File: test/performance/latency/LatencyTestPublisher.hpp [hunk 21]
```diff
@@ -135,6 +140,10 @@ class LatencyTestPublisher
     eprosima::fastdds::dds::DataReader* data_reader_;
     eprosima::fastdds::dds::DataReader* command_reader_;
 
+    /* QoS Profiles */
+    eprosima::fastdds::dds::DataReaderQos dr_qos;
+    eprosima::fastdds::dds::DataWriterQos dw_qos;
```
Comment (MiguelCompany) [inline test/performance/latency/LatencyTestPublisher.hpp:145]:
```suggestion
    eprosima::fastdds::dds::DataReaderQos dr_qos_;
    eprosima::fastdds::dds::DataWriterQos dw_qos_;
```

File: test/performance/latency/LatencyTestTypes.hpp (hunk 18 above)
Comment (JLBuenoLopez-eProsima) [reply test/performance/latency/LatencyTestTypes.hpp]: Done in 759afcbf9cb2c567b30b3b3475617648754d5895

Does the default constructor need to be modified? I do not think it is being used with the refactor.

File: test/performance/latency/LatencyTestTypes.cpp (hunk 19 above)
Comment (JLBuenoLopez-eProsima) [reply test/performance/latency/LatencyTestTypes.cpp]: Done in 759afcbf9cb2c567b30b3b3475617648754d5895

File: test/performance/latency/LatencyTestTypes.hpp (hunk 20 above)
Comment (JLBuenoLopez-eProsima) [reply test/performance/latency/LatencyTestTypes.hpp]: This method has not been included finally and instead the combination of delete/create is being used to resize the type. Consequently, this declaration has been erased in 759afcbf9cb2c567b30b3b3475617648754d5895.

File: test/performance/latency/LatencyTestPublisher.hpp (hunk 21 above)
Comment (JLBuenoLopez-eProsima) [reply test/performance/latency/LatencyTestPublisher.hpp]: Done in 759afcbf9cb2c567b30b3b3475617648754d5895

File: test/performance/latency/LatencyTestTypes.hpp (hunk 18 above)
Comment (MiguelCompany) [reply test/performance/latency/LatencyTestTypes.hpp]: No, it doesn't. It sets `m_typeSize = MAX_TYPE_SIZE` and `buffer_size_ = MAX_TYPE_SIZE - 4`, so alignment is ok.

**Comment by JLBuenoLopez-eProsima** [comment]
Rebase to master so BOUND_UNLIMITED can be used.

File: test/performance/latency/LatencyTestSubscriber.cpp [hunk 22]
```diff
@@ -461,50 +588,80 @@ bool LatencyTestSubscriber::test(
     }
     else
     {
-        latency_type_ = new LatencyType(datasize);
+        latency_type_ = static_cast<LatencyType*>(latency_data_type_->createData());
+        latency_type_->data.resize(datasize, 0);
```
[the remark this answers is not in the document]
Comment (MiguelBarro) [reply test/performance/latency/LatencyTestSubscriber.cpp]: If using the TypeSupport is the way to go we should refrain from keeping the old ways 😅

File: test/performance/latency/LatencyTestSubscriber.cpp (hunk 4 above)
Comment (MiguelBarro) [reply test/performance/latency/LatencyTestSubscriber.cpp]: If we create the data using TypeSupport we must stick to it in destruction to match the allocator used.

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
Partial review: data types

**Review by IkerLuengo — APPROVED** [review]
Looks good

**Review by IkerLuengo — APPROVED** [review]
:+1:

**Pull request merged**